### PR TITLE
feat: add legacy completions API support

### LINKED ISSUE #112: Support legacy completion models
Support ability to connect endpoints which implement legacy `/completions` [API](https://platform.openai.com/docs/api-reference/completions/create).

Make sure 
1. both AzureOpenAI and vanilla OpenAI authentication are supported,
2. streaming and non-streaming modes work

Success criteria: implement NF application via the adapter.

### PR DESCRIPTION
**Pull request by roman-romanov-o**
Closes https://github.com/epam/ai-dial-adapter-openai/issues/112

### CODE REVIEW

**Comment by ai-dial-actions** [comment]
Hey there and thank you for opening this pull request! 👋🏼

We require pull request titles to follow the [Conventional Commits specification](https://www.conventionalcommits.org/en/v1.0.0/) and it looks like your proposed title needs to be adjusted.

Details:

```
Unknown release type "Feature" found in pull request title "Feature: add legacy completions API support". 

Available types:
 - fix
 - hotfix
 - feat
 - feature
 - chore
```

**Comment by roman-romanov-o** [comment]
/deploy-review
>Environment URL: https://chat-ai-dial-adapter-openai-pr-113.nightly-test.deltixhub.io
>E2E tests status: failed

File: aidial_adapter_openai/legacy_completions.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+                },
+                "logprobs": choice["logprobs"],
+            }
+            for choice in item_dict["choices"]
+        ],
+    }
+
+
+async def legacy_completions(
+    data,
```
Comment (adubovik) [inline aidial_adapter_openai/legacy_completions.py]: Let's add type hints to the arguments

File: aidial_adapter_openai/legacy_completions.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+        "object": "chat.completion",
+        "id": item_dict["id"],
+        "model": item_dict["model"],
+        "created": item_dict["created"],
+        "usage": item_dict["usage"],
+        "choices": [
+            {
+                "index": choice["index"],
+                "finish_reason": choice["finish_reason"],
+                "message": {
```
Comment (adubovik) [inline aidial_adapter_openai/legacy_completions.py]: In the streaming mode the `message` field is called `delta`.

File: aidial_adapter_openai/app.py [hunk 3]
```diff
@@ -43,50 +51,10 @@
 init_telemetry(app, TelemetryConfig())
 configure_loggers()
 
-model_aliases: Dict[str, str] = json.loads(os.getenv("MODEL_ALIASES", "{}"))
```
Comment (adubovik) [inline aidial_adapter_openai/app.py:46]: Let's make the refactoring in a dedicated MR.

File: aidial_adapter_openai/app.py [hunk 4]
```diff
@@ -100,30 +68,40 @@ def get_api_version(request: Request):
 
 @app.post("/openai/deployments/{deployment_id}/chat/completions")
 async def chat_completion(deployment_id: str, request: Request):
+
     data = await parse_body(request)
 
     is_stream = data.get("stream", False)
 
-    api_type, api_key = await get_credentials(request, chat_completions_parser)
+    if deployment_id in LEGACY_COMPLETIONS_DEPLOYMENTS:
```
Comment (adubovik) [inline aidial_adapter_openai/app.py]: You really don't need to introduce `LEGACY_COMPLETIONS_DEPLOYMENTS` env var, because the format of the upstream endpoint already tells you if this is a chat completions model of a legacy completions model. The parsers do the work of distinguishing one from another.

File: aidial_adapter_openai/openai_override.py [hunk 5]
```diff
@@ -31,11 +35,90 @@ def handle_error_response_wrapper(wrapped, self, args, kwargs):
     raise OpenAIException(*args)
 
 
+# Overriden to handle cases, when server returns valid stream, but wihtout "text/event-stream" header
```
Comment (adubovik) [inline aidial_adapter_openai/openai_override.py]: wihtout => without

**Comment by roman-romanov-o** [comment]
/deploy-review
>Environment URL: https://chat-ai-dial-adapter-openai-pr-113.nightly-test.deltixhub.io
>E2E tests status: success

File: aidial_adapter_openai/legacy_completions.py (hunk 1 above)
Comment (roman-romanov-o) [reply aidial_adapter_openai/legacy_completions.py]: fixed, but adding type hints to data cause adding typing ignore somewhere, because linter thinks that

`data.get('n', 1)` still can be None

File: aidial_adapter_openai/legacy_completions.py (hunk 2 above)
Comment (roman-romanov-o) [reply aidial_adapter_openai/legacy_completions.py]: fixed

File: aidial_adapter_openai/app.py (hunk 4 above)
Comment (roman-romanov-o) [reply aidial_adapter_openai/app.py]: fixed

File: aidial_adapter_openai/openai_override.py (hunk 5 above)
Comment (roman-romanov-o) [reply aidial_adapter_openai/openai_override.py]: fixed

File: aidial_adapter_openai/legacy_completions.py [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+from aidial_adapter_openai.types import ChatCompletionRequestData
+from aidial_adapter_openai.utils.auth import get_auth_header
+from aidial_adapter_openai.utils.exceptions import handle_exceptions
+from aidial_adapter_openai.utils.remove_none import remove_none
+from aidial_adapter_openai.utils.sse_stream import to_openai_sse_stream
+from aidial_adapter_openai.utils.streaming import map_stream
+
+
+def convert_completion_to_chat_completion(
+    item: OpenAIObject, role="assistant", is_stream: bool = False
```
Comment (adubovik) [inline aidial_adapter_openai/legacy_completions.py]: The defaults here do not make much sense. Why non-streaming is more preferable default than streaming? They are both on the equal footing to me.
Same with the role parameter.

File: aidial_adapter_openai/legacy_completions.py [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+        "object": "chat.completion",
+        "id": item_dict["id"],
+        "model": item_dict["model"],
+        "created": item_dict["created"],
+        "usage": item_dict.get("usage") or {},
+        "choices": [
+            {
+                "index": choice["index"],
+                "finish_reason": choice["finish_reason"],
+                ("message" if is_stream else "delta"): {
```
Comment (adubovik) [inline aidial_adapter_openai/legacy_completions.py:36]: It's other way around. Consider reusing the helper:

https://github.com/epam/ai-dial-adapter-openai/blob/ce4a06b308123fa5d3d07ccbc87be4ee252d349d/aidial_adapter_openai/utils/streaming.py#L23-L45

File: aidial_adapter_openai/legacy_completions.py [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+    return {
+        "object": "chat.completion",
+        "id": item_dict["id"],
+        "model": item_dict["model"],
+        "created": item_dict["created"],
+        "usage": item_dict.get("usage") or {},
+        "choices": [
+            {
+                "index": choice["index"],
+                "finish_reason": choice["finish_reason"],
```
Comment (adubovik) [inline aidial_adapter_openai/legacy_completions.py:35]: Are you sure "finish_reason" will be present in every single chunk of response? Why not `choice.get("finish_reason")`?

File: aidial_adapter_openai/utils/remove_none.py [hunk 9]
```diff
@@ -0,0 +1,11 @@
+from typing import Any, Dict, Optional
+
+
+def remove_none(d: Dict[str, Optional[Any]]) -> Dict[str, Any]:
+    """
```
Comment (adubovik) [inline aidial_adapter_openai/utils/remove_none.py:5]: IMO, 6 lines of comments for 1 line of actual code is an overkill.

Noone ever is going to read this comment from start to finish. 
One would jump straight away to the line of code.

File: aidial_adapter_openai/legacy_completions.py [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+
+async def legacy_completions(
+    data: ChatCompletionRequestData,
+    deployment_id: str,
+    is_stream: bool,
+    api_key: str,
+    api_type: str,
+    api_version: str,
+    api_base: str,
+    model=None,
```
Comment (adubovik) [inline aidial_adapter_openai/legacy_completions.py]:
```
model: str | None = None,
engine: str | None = None
```

File: aidial_adapter_openai/utils/parsers.py [hunk 11]
```diff
@@ -53,22 +58,32 @@ def parse(self, endpoint: str) -> AzureOpenAIEndpoint | OpenAIEndpoint:
             )
 
         match = re.search(f"(.+?)/{self.name}", endpoint)
-
         if match:
             return OpenAIEndpoint(api_base=match[1])
+        return match
+
+    def is_valid(self, endpoint: str) -> bool:
+        if self._find_match(endpoint):
```
Comment (adubovik) [inline aidial_adapter_openai/utils/parsers.py:66]: `return self._find_match(endpoint)`

File: aidial_adapter_openai/utils/parsers.py [hunk 12]
```diff
@@ -53,22 +58,32 @@ def parse(self, endpoint: str) -> AzureOpenAIEndpoint | OpenAIEndpoint:
...
+            return True
+        return False
+
+    def parse(self, endpoint: str) -> AzureOpenAIEndpoint | OpenAIEndpoint:
+        if match := self._find_match(endpoint):
```
Comment (adubovik) [inline aidial_adapter_openai/utils/parsers.py:69]: `return self._find_match(endpoint)`

File: aidial_adapter_openai/utils/parsers.py (hunk 11 above)
Comment (adubovik) [inline aidial_adapter_openai/utils/parsers.py]: `return self._find_match(endpoint) is not None`

File: aidial_adapter_openai/app.py [hunk 13]
```diff
@@ -100,15 +67,37 @@ def get_api_version(request: Request):
...
     data = await parse_body(request)
+    is_stream = bool(data.get("stream", False))
+    api_version = get_api_version(request)
 
-    is_stream = data.get("stream", False)
+    upstream_endpoint = request.headers["X-UPSTREAM-ENDPOINT"]
 
-    api_type, api_key = await get_credentials(request, chat_completions_parser)
+    # If it is not chat completions, but legacy completions
+    if not chat_completions_parser.is_valid(
```
Comment (adubovik) [inline aidial_adapter_openai/app.py:78]: It's better to parse and validate at the same time, so that you don't need to keep in mind that every `parse` should be only called under the branch where `is_valid` is True:

```python
if (endpoint := completions_parser.parse(upstream_endpoint)) is not None:
    ...
```

File: aidial_adapter_openai/legacy_completions.py (hunk 1 above)
Comment (adubovik) [reply aidial_adapter_openai/legacy_completions.py]: `data: dict` is enough :)

I want all of the parameters to be typed to a reasonable extent, which is the case for the rest of the project.

Comment (roman-romanov-o) [reply aidial_adapter_openai/legacy_completions.py]: I think, that such `data: dict` type hints are mostly useless, other than preventing some really basic errors.
It's the same as adding `data: object` or `data: Any`.

Even when I know that data is `dict/list/object`, I have no idea at all, how to actually work with this parameter inside of code.

Comment (roman-romanov-o) [reply aidial_adapter_openai/legacy_completions.py]: I understand that it's hard to add specific types to all nested objects, but at least we can do it in some places, to make life of developer easier, and prevent him from making `print(data)` to actually understand, what's hidden inside of this dictionary

File: aidial_adapter_openai/utils/remove_none.py (hunk 9 above)
Comment (roman-romanov-o) [reply aidial_adapter_openai/utils/remove_none.py:5]: Currently it doesn't make sense, you're right

But it could make sense to add such `reStructuredText` (or maybe google/numpy type of comments) to each function, so at some point we could generate automatic documentation using some tool like `Sphinx`

File: aidial_adapter_openai/utils/parsers.py (hunk 12 above)
Comment (roman-romanov-o) [reply aidial_adapter_openai/utils/parsers.py:69]: `self._find_match(endpoint)`could be None, need to check it before returning

File: aidial_adapter_openai/utils/parsers.py (hunk 11 above)
Comment (roman-romanov-o) [reply aidial_adapter_openai/utils/parsers.py]: Done

File: aidial_adapter_openai/app.py (hunk 13 above)
Comment (roman-romanov-o) [reply aidial_adapter_openai/app.py:78]: This would not work with current parsers, since they are raising `DialException` if parsing has not succeeded
I don't really want to change behaviour of this class 

---

Also, another problem, that raises when we send both `/chat/completions` and `/completions` to parser, is that with current `EndpointParser`,

`EndpointParser(name="completions").parse("/some_deployment/chat/completions") will parse successfully, with
`deployment_id = some_deployment/chat`

Comment (roman-romanov-o) [reply aidial_adapter_openai/app.py:78]: The approach here, other than I use right now (check that chat completions is not valid in same statement, to filter chat completions from completions) is to create subclass

```python
class CompletionsEndpointParser(EndpointParser):
    ...
```
to exclude `chat/completions` case

File: aidial_adapter_openai/legacy_completions.py (hunk 6 above)
Comment (roman-romanov-o) [reply aidial_adapter_openai/legacy_completions.py]: fixed, removed defaults

File: aidial_adapter_openai/utils/remove_none.py (hunk 9 above)
Comment (adubovik) [reply aidial_adapter_openai/utils/remove_none.py:5]: I agree, but we don't have such plans yet. 

Also, automatic documentation is a good idea for libraries, but this adapter isn't a library but an application on its own.

File: aidial_adapter_openai/app.py (hunk 13 above)
Comment (adubovik) [reply aidial_adapter_openai/app.py:78]: > This would not work with current parsers

The parsers aren't set in stone.

This condition looks so unnatural, because we didn't bother to fix the parsers when `/completions` case was introduced. Conceptually parsers should be disjoint.
